### PR TITLE
[SE-0101] Migration support fix-it

### PR DESCRIPTION
**Pull request by rintaro**
#### What's in this pull request?

Split from #3854
Implement migration support fix-it for SE-0101

```
error: 'sizeof' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeof(C.self)
      ^~~~~~~ ~~~~~~
      MemoryLayout< >.size
```

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by gribozavr** [comment]
@rintaro Thank you!

File: test/expr/expressions.swift [hunk 1]
```diff
@@ -872,3 +872,27 @@ let _ = (x, 3).1
...
+  _ = sizeof(P.Value.self) // expected-error {{'sizeof' is unavailable: use MemoryLayout<T>.size instead.}} {{7-14=MemoryLayout<}} {{21-27=>.size}} {{none}}
+  _ = sizeof(Cse0101<P>.self) // expected-error {{'sizeof' is unavailable: use MemoryLayout<T>.size instead.}} {{7-14=MemoryLayout<}} {{24-30=>.size}} {{none}}
+  _ = alignof(Cse0101<P>.T.self) // expected-error {{'alignof' is unavailable: use MemoryLayout<T>.alignment instead.}} {{7-15=MemoryLayout<}} {{27-33=>.alignment}} {{none}}
+  _ = strideof(P.Type.self) // expected-error {{'strideof' is unavailable: use MemoryLayout<T>.stride instead.}} {{7-16=MemoryLayout<}} {{22-28=>.stride}} {{none}}
+
+  _ = sizeofValue(x) // expected-error {{'sizeofValue' is unavailable: use MemoryLayout<T>.size instead.}} {{7-21=MemoryLayout<Cse0101<P>>.size}} {{none}}
+  _ = alignofValue(x.val) // expected-error {{'alignofValue' is unavailable: use MemoryLayout<T>.alignment instead.}} {{7-26=MemoryLayout<P>.alignment}} {{none}}
+  _ = strideofValue(x.val.getIt()) // expected-error {{'strideofValue' is unavailable: use MemoryLayout<T>.stride instead.}} {{7-35=MemoryLayout<P.Value>.stride}} {{none}}
```
Comment (rintaro) [inline test/expr/expressions.swift]: `strideofValue(x.val.getIt())` => `MemoryLayout<P.Value>.stride`
Do we need preserve `x.val.getIt()`?
i.e. `MemoryLayout<P.Value>.stride; _ = x.val.getIt()`

**Comment by rintaro** [comment]
One more TODO: we need special treatment for `let x: Int = 1; sizeof(x.dynamicType)`
We can't `MemoryLayout<x.dynamicType>.size`.
Since that is equivalent to `sizeofValue(x)`, we should replace it to `MemoryLayout<Int>.size`

**Comment by rintaro** [comment]
```swift
func _func<C: Collection>(c: C)  where C.Index == Int {
  _ = sizeof(C)
  _ = sizeof(C.self)
  _ = sizeof(C.Iterator.self)
  _ = sizeof(c.dynamicType)
  _ = sizeofValue(c)
  _ = sizeofValue(c[0])
}
```
result
```swift
test0101.swift:2:14: warning: missing '.self' for reference to metatype of type 'C'
  _ = sizeof(C)
             ^
              .self
test0101.swift:2:7: error: 'sizeof' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeof(C)
      ^~~~~~~ ~
      MemoryLayout< >.size
test0101.swift:3:7: error: 'sizeof' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeof(C.self)
      ^~~~~~~ ~~~~~~
      MemoryLayout< >.size
test0101.swift:4:7: error: 'sizeof' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeof(C.Iterator.self)
      ^~~~~~~          ~~~~~~
      MemoryLayout<    >.size
test0101.swift:5:7: error: 'sizeof' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeof(c.dynamicType)
      ^~~~~~~~~~~~~~~~~~~~~
      MemoryLayout<C>.size
test0101.swift:6:7: error: 'sizeofValue' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeofValue(c)
      ^~~~~~~~~~~~~~
      MemoryLayout<C>.size
test0101.swift:7:7: error: 'sizeofValue' is unavailable: use MemoryLayout<T>.size instead.
  _ = sizeofValue(c[0])
      ^~~~~~~~~~~~~~~~~
      MemoryLayout<C.Iterator.Element>.size
```

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
```
/home/buildnode/jenkins/workspace/swift-PR-Linux/swift/test/expr/expressions.swift:894:15: error: unexpected warning produced: '.dynamicType' is deprecated. Use 'type(of: ...)' instead
  _ = sizeof(x.dynamicType) // expected-error {{'sizeof' is unavailable: use MemoryLayout<T>.size instead.}} {{7-28=MemoryLayout<Cse0101<P>>.size}} {{none}}
              ^
```
Need to rebase on #3878 (and  #3885).

**Comment by gribozavr** [comment]
@rintaro Would you mind rebasing?  Thanks!

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by gribozavr** [comment]
@rudkx Would you mind reviewing the updated migration code?

**Comment by rudkx** [comment]
LGTM!

**Pull request merged**